File: 143-ReorderList/143-ReorderList.py
# Last updated: 8/13/2025, 8:22:08 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # get the middle point

        slow = head
        fast = head

        while fast is not None and fast.next is not None : 
            slow = slow.next
            fast = fast.next.next

        second_head = slow.next
        slow.next = None #dividing the two list

        #reversing the second list
        prev = None
        curr = second_head

        while curr is not None : 
            curr_next = curr.next
            curr.next = prev
            prev = curr
            curr = curr_next
        
        #adding to the list one by one
        p1 = head
        p2 = prev

        i = 1

        while p1 is not None and p2 is not None : 
            if i%2 == 0 : 
                p2_next = p2.next
                p1_next = p1.next
                p1.next = p2
                p2.next = p1_next
                p2 = p2_next
                p1 = p1_next
            i+=1
        
        return head

        
        

File: 143-ReorderList/143-ReorderList.java
// Last updated: 8/13/2025, 8:22:14 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {

        //finding the middle of the linkedlist
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next!=null && fast.next.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode headSecond = slow.next;
        slow.next = null;

        //reversing the second part
        ListNode prev = null;
        ListNode curr = headSecond;
        while(curr!=null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        headSecond = prev;
        //slow.next= headSecond;

        //connecting the linked list
        ListNode headFirst = head;
        while(headSecond!=null){
            ListNode tempFirst = headFirst.next;
            ListNode tempSecond = headSecond.next;
            headFirst.next = headSecond;
            headSecond.next = tempFirst;
            headSecond = tempSecond;
            headFirst = tempFirst;
        }

        
    }
}